etcdctl/ctlv3/command/perf_validate.go

```diff
@@ -0,0 +1,185 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package command
+
+import (
+	"context"
+	"encoding/binary"
+	"fmt"
+	"math"
+	"math/rand"
+	"sync"
+	"time"
+
+	v3 "github.com/coreos/etcd/clientv3"
+	"github.com/coreos/etcd/pkg/report"
+
+	"github.com/spf13/cobra"
+	"golang.org/x/time/rate"
+	"gopkg.in/cheggaaa/pb.v1"
+)
+
+var (
+	validPerfLoad string
+)
+
+type validatePerfCfg struct {
+	limit    int
+	clients  int
+	duration int
+}
+
+var validatePerfCfgMap = map[string]validatePerfCfg{
+	// TODO: support read limit
+	"s": validatePerfCfg{
+		limit:    150,
+		clients:  50,
+		duration: 60,
+	},
+	"m": validatePerfCfg{
+		limit:    1000,
+		clients:  200,
+		duration: 60,
+	},
+	"l": validatePerfCfg{
+		limit:    8000,
+		clients:  500,
+		duration: 60,
+	},
+	"xl": validatePerfCfg{
+		limit:    15000,
+		clients:  1000,
+		duration: 60,
+	},
+}
+
+// NewValidatePerfCommand returns the cobra command for "validate-perf".
+func NewValidatePerfCommand() *cobra.Command {
+	cmd := &cobra.Command{
+		Use:   "validate-perf [options]",
+		Short: "Validate the performance of the etcd cluster",
+		Run:   newValidatePerfCommand,
+	}
+
+	// TODO: support customized configuration
+	cmd.Flags().StringVar(&validPerfLoad, "load", "s", "s(small), m(medium), l(large), xl(xLarge)")
+	return cmd
+}
+
+// newValidatePerfCommand executes the "validate-perf" command.
+func newValidatePerfCommand(cmd *cobra.Command, args []string) {
+	var cfg validatePerfCfg
+	switch validPerfLoad {
+	case "s", "small":
+		cfg = validatePerfCfgMap["s"]
+	case "m", "medium":
+		cfg = validatePerfCfgMap["m"]
+	case "l", "large":
+		cfg = validatePerfCfgMap["l"]
+	case "xl", "xLarge":
+		cfg = validatePerfCfgMap["xl"]
+	default:
+		ExitWithError(ExitBadFeature, fmt.Errorf("unkwon load option %v", validPerfLoad))
+	}
+
+	requests := make(chan v3.Op, cfg.clients)
+	limit := rate.NewLimiter(rate.Limit(cfg.limit), 1)
+
+	var clients []*v3.Client
+	for i := 0; i < cfg.clients; i++ {
+		clients = append(clients, mustClientFromCmd(cmd))
+	}
+
+	ksize, vsize := 256, 1024
+	k, v := make([]byte, ksize), string(make([]byte, vsize))
+
+	bar := pb.New(cfg.duration)
+	bar.Format("Bom !")
+	bar.Start()
+
+	r := report.NewReport("%4.4f")
+	var wg sync.WaitGroup
+
+	for i := range clients {
+		wg.Add(1)
+		go func(c *v3.Client) {
+			defer wg.Done()
+			for op := range requests {
+				st := time.Now()
+				_, err := c.Do(context.Background(), op)
+				r.Results() <- report.Result{Err: err, Start: st, End: time.Now()}
+			}
+		}(clients[i])
+	}
+
+	go func() {
+		ctx, _ := context.WithTimeout(context.Background(), time.Duration(cfg.duration)*time.Second)
+
+		for limit.Wait(ctx) == nil {
+			binary.PutVarint(k, int64(rand.Intn(math.MaxInt64)))
+			requests <- v3.OpPut(string(k), v)
```

Comment:
Good idea. My initial thought is that this tool should be used pre installation. But we can make it better with very little effort.